diff --git a/sipsimple/account/xcap/__init__.py b/sipsimple/account/xcap/__init__.py
index a015c13d..3c088694 100644
--- a/sipsimple/account/xcap/__init__.py
+++ b/sipsimple/account/xcap/__init__.py
@@ -10,7 +10,7 @@ import os
 import random
 import socket
 import weakref
-import gevent
+#import gevent

 from io import StringIO
 from collections import OrderedDict
@@ -1844,8 +1844,8 @@ class XCAPManager(object):
         NotificationCenter().post_notification('XCAPManagerDidReloadData', sender=self, data=data)

     def _fetch_documents(self, documents):
-        jobs = [gevent.spawn(document.fetch) for document in (doc for doc in self.documents if doc.name in documents and doc.supported)]
-        gevent.joinall(jobs, timeout=15)
+        # jobs = [gevent.spawn(document.fetch) for document in (doc for doc in self.documents if doc.name in documents and doc.supported)]
+        # gevent.joinall(jobs, timeout=15)

     def _save_journal(self):
         try:

