diff -ruN pjsip/Makefile pjsip-patch/Makefile
--- pjsip/Makefile	2020-02-14 10:48:27.000000000 +0100
+++ pjsip-patch/Makefile	2024-09-03 10:32:23.926160100 +0200
@@ -129,10 +129,9 @@
 		sed -e "s!@INCLUDEDIR@!$(includedir)!" | \
 		sed -e "s!@LIBDIR@!$(libdir)!" | \
 		sed -e "s/@PJ_VERSION@/$(PJ_VERSION)/" | \
-		sed -e "s!@PJ_LDLIBS@!!" | \
-		sed -e "s!@PJ_LDXXLIBS@!$(PJ_LDXXLIBS)!" | \
-		sed -e "s!@PJ_INSTALL_CFLAGS@!!" | \
-		sed -e "s!@PJ_INSTALL_CXXFLAGS@!$(PJ_INSTALL_CXXFLAGS)!" > $(DESTDIR)/$(libdir)/pkgconfig/libpjproject.pc
+		sed -e "s!@PJ_INSTALL_LDFLAGS@!$(PJ_INSTALL_LDFLAGS)!" | \
+		sed -e "s!@PJ_INSTALL_LDFLAGS_PRIVATE@!$(PJ_INSTALL_LDFLAGS_PRIVATE)!" | \
+		sed -e "s!@PJ_INSTALL_CFLAGS@!$(PJ_INSTALL_CFLAGS)!" > $(DESTDIR)$(libdir)/pkgconfig/libpjproject.pc
 
 uninstall:
 	$(RM) $(DESTDIR)$(libdir)/pkgconfig/libpjproject.pc
diff -ruN pjsip/aconfigure pjsip-patch/aconfigure
--- pjsip/aconfigure	2024-09-03 10:10:13.926165600 +0200
+++ pjsip-patch/aconfigure	2024-09-03 10:32:23.941785600 +0200
@@ -5156,6 +5156,12 @@
 
 	$as_echo "#define WIN32_LEAN_AND_MEAN 1" >>confdefs.h
 
+	case $target in
+	    *_64-w64-mingw* )
+		$as_echo "#define PJ_WIN64 1" >>confdefs.h
+
+	    ;;
+	esac
 	;;
     *darwin*)
 	$as_echo "#define PJ_DARWINOS 1" >>confdefs.h
diff -ruN pjsip/aconfigure.ac pjsip-patch/aconfigure.ac
--- pjsip/aconfigure.ac	2020-02-14 10:48:27.000000000 +0100
+++ pjsip-patch/aconfigure.ac	2024-09-03 10:32:23.973036300 +0200
@@ -161,6 +161,11 @@
 	AC_DEFINE(PJ_WIN32,1)
 	AC_DEFINE(PJ_WIN32_WINNT,0x0400)
 	AC_DEFINE(WIN32_LEAN_AND_MEAN)
+	case $target in
+	    *_64-w64-mingw* )
+		AC_DEFINE(PJ_WIN64,1)
+	    ;;
+	esac
 	;;
     *darwin*)
 	AC_DEFINE(PJ_DARWINOS,1)
diff -ruN pjsip/pjlib/include/pj/compat/os_auto.h.in pjsip-patch/pjlib/include/pj/compat/os_auto.h.in
--- pjsip/pjlib/include/pj/compat/os_auto.h.in	2020-02-14 10:48:27.000000000 +0100
+++ pjsip-patch/pjlib/include/pj/compat/os_auto.h.in	2024-09-03 10:32:23.988659900 +0200
@@ -30,6 +30,7 @@
 #undef PJ_OS_NAME
 
 /* Legacy macros */
+#undef PJ_WIN64
 #undef PJ_WIN32
 #undef PJ_WIN32_WINNT
 #undef WIN32_LEAN_AND_MEAN
diff -ruN pjsip/pjnath/src/pjnath-test/stun_sock_test.c pjsip-patch/pjnath/src/pjnath-test/stun_sock_test.c
--- pjsip/pjnath/src/pjnath-test/stun_sock_test.c	2020-02-14 10:48:27.000000000 +0100
+++ pjsip-patch/pjnath/src/pjnath-test/stun_sock_test.c	2024-09-03 10:32:23.988659900 +0200
@@ -25,7 +25,6 @@
     RESPOND_STUN    = 1,
     WITH_MAPPED	    = 2,
     WITH_XOR_MAPPED = 4,
-
     ECHO	    = 8
 };
 
