--- pjproject-2.10/pjmedia/src/pjmedia/ffmpeg_util.h	2021-03-19 16:51:27.589562664 +0100
+++ pjsip/pjmedia/src/pjmedia/ffmpeg_util.h	2021-03-19 17:00:32.821251088 +0100
@@ -38,13 +38,13 @@
 #include <libavcodec/avcodec.h>
 
 
-#ifdef PJMEDIA_USE_OLD_FFMPEG
-#   define AVPixelFormat	PixelFormat
-#   define AV(str)		str
-#   define PIX_FMT_GBRP 	PIX_FMT_GBR24P
-#else
+//#ifdef PJMEDIA_USE_OLD_FFMPEG
+//#   define AVPixelFormat	PixelFormat
+//#   define AV(str)		str
+//#   define PIX_FMT_GBRP 	PIX_FMT_GBR24P
+//#else
 #   define AV(str)		AV_ ## str
-#endif
+//#endif
 					 
 
 void pjmedia_ffmpeg_add_ref();
diff -ruN pjproject-2.10/pjmedia/src/pjmedia/ffmpeg_util.c pjsip/pjmedia/src/pjmedia/ffmpeg_util.c
--- pjproject-2.10/pjmedia/src/pjmedia/ffmpeg_util.c	2020-02-14 10:48:27.000000000 +0100
+++ pjsip/pjmedia/src/pjmedia/ffmpeg_util.c	2021-02-06 20:25:00.472499296 +0100
@@ -34,10 +34,11 @@
     enum AVPixelFormat	pf;
 } ffmpeg_fmt_table[] =
 {
+    { PJMEDIA_FORMAT_ARGB, AV(PIX_FMT_ARGB)},
     { PJMEDIA_FORMAT_RGBA, AV(PIX_FMT_RGBA)},
     { PJMEDIA_FORMAT_RGB24,AV(PIX_FMT_BGR24)},
     { PJMEDIA_FORMAT_BGRA, AV(PIX_FMT_BGRA)},
-    { PJMEDIA_FORMAT_GBRP, AV(PIX_FMT_GBRP)},
+    { PJMEDIA_FORMAT_GBRP, AV(PIX_FMT_GBR24P)},
 
     { PJMEDIA_FORMAT_AYUV, AV(PIX_FMT_NONE)},
     { PJMEDIA_FORMAT_YUY2, AV(PIX_FMT_YUYV422)},
